Add task solution

agnijaravdo opened the pull request.

agnijaravdo committed:
add components with solution

agnijaravdo committed:
add snapshot test for component

agnijaravdo committed:
Delete package-lock.json

agnijaravdo committed:
Revert "Delete package-lock.json"

This reverts commit 0b0495942d81a505b2927c55fdb571b001954fbf.

agnijaravdo committed:
Update package-lock.json

nerijus-muranovas commented on src/App.css:
```diff
@@ -1,3 +1,12 @@
-}
\ No newline at end of file
+}
+label {
```
Use classes instead of overriding native elements
```suggestion
.some-label {
```

nerijus-muranovas commented on src/App.css:
```diff
@@ -1,3 +1,12 @@
+  text-align: center;
+}
+
+.attendantsDisplay span {
```
Same as above. One concrete class would be enough e.g.
```suggestion
.attendants-title {
```

nerijus-muranovas commented on src/components/AttendantDisplay.jsx:
```diff
@@ -0,0 +1,14 @@
+      <br />
+    </div>
+  );
+}
```
The practice is to define this as a `const SomeComponent` and then export it as default at the bottom of the file.
Also the name can be simplified.
```suggestion
const Attendant = ({
  attendant: { name, lastName, jobTitle, age },
}) => (
    <div className="attendantsDisplay">
      <br />
      Name: <span>{name} </span>
      Last Name: <span>{lastName} </span>
      Job Title: <span>{jobTitle} </span>
      Age: <span>{age} </span>
      <br />
    </div>
  );
  
  export default Attendant
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+  }
+
+  function clearForm() {
+    [setName, setLastName, setJobTitle, setAge].forEach((setter) => setter(""));
```
Interesting approach. This or just setting each state to empty string is good

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+  useEffect(() => {
+    setIsLoading(true);
+
+    Promise.all([getJobTitles(), getAttendants()])
```
Interesting approach. Has two drawbacks:
1. let's say job titles are fetched faster while attendants are not. Since these parts are not dependant on each other, it would be good to have them operate on their own;
2. the code inside `.then` block handles both promises, which would make error handling more difficult. Let's say job titles fetch fails - then attendants would fail too.

The topic for this is [separation of concerns](https://deviq.com/principles/separation-of-concerns).

A more optimal solution is to have both calls to api in their separate functions (also using separate loading states).

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+        }
+      })
+      .catch((e) => {
+        console.error("failed to get api response", e);
```
The error can be stored in an error state and displayed. Also logging to console should not be done.

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+      .catch((e) => {
+        console.error("failed to get api response", e);
+      })
+      .finally(() => {
```
The more common practice is for `then`, `catch`, `final` to be written in an `async` function using a `try catch` block. Codedamn has a [very nice example](https://codedamn.com/news/javascript/javascript-async-await-error#:~:text=async%2Dawait%20code.-,Using%20try%2Dcatch,-The%20most%20common).

Edit: just checked - your submit function is a very good example 👍

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+
+  async function submitHandler(e) {
+    e.preventDefault();
+    const submittedData = {
```
To be more specific:
```suggestion
    const newAttendant = {
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+    try {
+      setIsLoading(true);
+      const { status } = await addAttendant(submittedData);
+      if (status === 200) {
```
Reversing the function for early return makes the code more clean. javascript.plainenglish.io has a [good example](https://javascript.plainenglish.io/early-return-with-react-hooks-f96fa4a33124#:~:text=What%20is%20Early%20Return%3F).

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+      if (status === 200) {
+        const { status: attendancesStatus, data } = await getAttendants();
+        if (attendancesStatus === 200) {
+          setAttendants([...data].sort((a, b) => a.age - b.age));
```
The array spreading can be avoided since the data is already an array.
```suggestion
          setAttendants(data.sort((a, b) => a.age - b.age));
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+  const [allJobTitles, setAllJobTitles] = useState([]);
+  const [isLoading, setIsLoading] = useState(false);
+
+  async function submitHandler(e) {
```
Functions that are passed to components usually are named `handleSomething`. In this case it would be:
```suggestion
  async function handleSubmit(e) {
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+      clearForm();
+      setIsLoading(false);
+    }
+  }
```
For this submit handler apply the same principle as for the `useEffect` below - for each call have their separate functions for easier readability.

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+import React, { useState, useEffect } from "react";
```
You could improve `AttendantForm` by:
- fetching attendants data on the page level (`App` component) and displaying it as a separate component. This would have the concerns of form separate from concerns of reading attendants data.
- fetching `jobTitles` on the page level (`App` component) and passing `jobTitles` as a prop to this component.
- having an `onSubmit` prop function that would be called when `submit` button is pressed (in the submitHandler function).

By doing the above, this component would only have the concerns of getting the inputted new attendant data and passing it to the parent component for uploading to the api.

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+  const [lastName, setLastName] = useState("");
+  const [jobTitle, setJobTitle] = useState("");
+  const [age, setAge] = useState("");
+  const [allJobTitles, setAllJobTitles] = useState([]);
```
Can be shortened to
```suggestion
  const [jobTitles, setJobTitles] = useState([]);
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+        }
+      }
+    } catch (e) {
+      console.error("failed to add attendant", e);
```
Same for error handling as mentioned in the `useEffect` below.

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+          key={`${e.name}${e.lastName}${e.jobTitle}${e.age}`}
+          attendant={e}
+        />
+      ))}
```
Since attendants do not have ids, an index can be used and shortened. Also the mapping function can be extracted to a `renderAttendants` function before the `return`.
```suggestion
      {attendants.map((attendant, index) => (
        <AttendantDisplay
          key={`attendant-${index}`}
          attendant={attendant}
        />
      ))}
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+      <h1>Attendant Registration Form</h1>
+      <br />
+      <form
+        className="attendant-form"
```
Seems like there is no `attendant-form` class. Unused classes can be removed for extra tidiness

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+            required
+            onChange={(e) => {
+              setName(e.target.value);
+            }}
```
The practice is to have a separate `handleSomethingChange` function that is passed to the `onChange` prop. Same goes for all html elements below
```suggestion
            onChange={handleNameChange}
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+            onInvalid={(e) =>
+              e.target.setCustomValidity("Please enter a valid name")
+            }
+            onInput={(e) => e.target.setCustomValidity("")}
```
Interesting approach to these. The more common (and maintainable) practice is to have a separate error state for each input. It would be more maintainable that way (since the inputs would hook into the react lifecycle).

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+          >
+            <option
+              value=""
+              style={{ textAlign: "center" }}
```
Better to use `className` instead of inline styles according to [react developers reccomendations](https://legacy.reactjs.org/docs/dom-elements.html#style).

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+            >
+              {isLoading
+                ? "Loading job titles... 🔄"
+                : "Please select a job title ⬇️"}
```
It's a nice touch to have this decorative option and it would be even better if instead of the whole select element, loading text would be displayed. I do not recommend having an option with an empty value.

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,191 @@
+              >
+                {jobTitle}
+              </option>
+            ))}
```
For mapping function - same feedback as for the attendants comment below.

nerijus-muranovas reviewed (commented):
There are a few places to work on but in general - good effort 👍

agnijaravdo committed:
address suggested changes 1/2

TODO: refactor AttendantForm validations to not depend on HTML5 form validations

agnijaravdo commented:
still left to do:
- refactor AttendantForm validations to not depend on HTML5 form validations. Will add custom validation methods
- add unit tests

agnijaravdo committed:
add custom validations instead of HTML5 validations

agnijaravdo committed:
add aria attributes to improve accessibility

agnijaravdo committed:
improve styling/UI

agnijaravdo committed:
add unit test for AttendantForm component

nerijus-muranovas commented on src/App.js:
```diff
@@ -1,10 +1,90 @@
+      const { status } = await addAttendant(newAttendant);
+      if (status !== 200) {
+        return;
+      }
```
This code block does not really do anything so it's ok to remove it

nerijus-muranovas commented on src/App.js:
```diff
@@ -1,10 +1,90 @@
+    };
+
+    handleAddAttendant(newAttendant);
+    handleFetchAttendants();
```
Since these are asynchronous, it would really help to `await` both of them so `handleFetchAttendants` would only always happen after `handleAddAttendant`

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,320 @@
+    [setName, setLastName, setJobTitle, setAge].forEach((setter) => setter(""));
+  };
+
+  const renderAttendants = () =>
```
I suggest rendering attendants in the App.js file so it would be a tad cleaner

agnijaravdo committed:
add Attendants and App unit tests with mocked API calls

nerijus-muranovas commented on src/App.test.js:
```diff
@@ -1,8 +1,60 @@
+  it("renders component correctly on successful initial load", async () => {
+    const { container } = render(<App />);
+
+    await waitFor(() => {
+      expect(screen.queryByTestId("job-titles-loader")).not.toBeInTheDocument();
+    });
```
Instead of `waitFor queryBy`, you can use `await findBy`. Does the same with less code

nerijus-muranovas commented on src/App.test.js:
```diff
@@ -1,8 +1,60 @@
+    });
+  });
+
+  it("renders component correctly on successful initial load", async () => {
```
Company practice is to use `when`
```suggestion
  it("renders component correctly when data is loaded", async () => {
```

nerijus-muranovas commented on src/App.test.js:
```diff
@@ -1,8 +1,60 @@
+
+  it("renders title", async () => {
+    render(<App />);
+    await waitFor(async () => {
```
This test would probably work without the `waitFor`

agnijaravdo replied:
without using waitFor test returns  `An update to App inside a test was not wrapped in act(...).` 😕

nerijus-muranovas commented on src/components/Attendant.test.jsx:
```diff
@@ -0,0 +1,32 @@
+    expect(container).toMatchSnapshot();
+  });
+
+  it("renders component with given props", async () => {
```
Since this component only renders data from the props, a snapshot test is enough

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,320 @@
+  const handleChange = (e) => {
+    const { name, value } = e.target;
+    switch (name) {
+      case "name":
+        setName(value);
+        validateNameInput(value);
+        break;
+      case "lastName":
+        setLastName(value);
+        validateLastNameInput(value);
+        break;
+      case "jobTitle":
+        setJobTitle(value);
+        validateJobTitleSelection(value);
+        break;
+      case "age":
+        setAge(value);
+        validateAgeInput(value);
+        break;
```
One validation function for the whole new attendant object would be awesome

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,320 @@
+    }
+  };
+
+  const handleBlur = (e) => {
```
This and `handleChange` functions seem to do almost the same. I suggest having a separate function that would handle the validation for both `handleBlur` and `handleChange`.

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,320 @@
+    }
+  };
+
+  const validateNameInput = (value) => {
```
These validation functions could be placed in a separate file

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,320 @@
+  attendantsApiError,
+  jobTitlesApiError,
+}) => {
+  const [name, setName] = useState("");
+  const [lastName, setLastName] = useState("");
+  const [jobTitle, setJobTitle] = useState("");
+  const [age, setAge] = useState("");
```
You could try to store all form data in a [single state object](https://stackoverflow.com/questions/64754900/reactjs-using-object-to-store-form-data-as-state-variable). That would make the code more DRY

agnijaravdo committed:
fix according to suggested changes

nerijus-muranovas commented on src/App.js:
```diff
@@ -1,10 +1,109 @@
+        key={`attendant-${index}`}
+        attendant={attendant}
+      />
+    ))}
```
Some of these lines look misaligned. I suggest downloading `prettier` extension and enabling `format on save` in the settings.

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,273 @@
+
+  async function handleSubmit(e) {
+    e.preventDefault();
+    await submitAttendant({ ...formInputs });
```
```suggestion
    await submitAttendant(formInputs);
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,273 @@
+    } else {
+      setFormErrors({ ...formErrors, ageError: "" });
+    }
+  }
```
For the field change handlers, I bet you could make an abstraction for them in a single function.

The function could look something like:
```
const handleFieldChange = (event, field, errorKey) => { ... }
```

nerijus-muranovas commented on src/components/AttendantForm.jsx:
```diff
@@ -0,0 +1,273 @@
+    if (!validateInputs({name: "age", value})) {
+      handleValidationError("age");
+    } else {
+      setFormErrors({ ...formErrors, ageError: "" });
```
If new state is dependant on old state [use a callback](https://react.dev/reference/react/useState#setstate).

agnijaravdo replied:
According to ["Is using an updater always preferred? "](https://react.dev/reference/react/useState#is-using-an-updater-always-preferred) this looks to be not the case where updater function is needed. Is it your personal preference to always use updater function or am I missing something?

nerijus-muranovas commented on src/utils/attendantFormInputValidations.js:
```diff
@@ -0,0 +1,38 @@
+  return !!value.match(ageRegex);
+}
+
+const validateInputs = ({name, value}) => {
```
This function can be exported instead of the method used below
```suggestion
export const validateInputs = ({name, value}) => {
```

nerijus-muranovas commented on src/components/AttendantForm.test.jsx:
```diff
@@ -0,0 +1,151 @@
+const jobTitles = ["Accountant", "Engineer", "Lawyer", "Teacher"];
+
+const fillAndSubmitForm = async () => {
+  userEvent.type(screen.getByTestId("name"), "John");
+  userEvent.type(screen.getByTestId("last-name"), "Doe");
+  userEvent.selectOptions(await screen.findByTestId("job-title"), "Engineer");
+  userEvent.type(screen.getByTestId("age"), "40");
+  userEvent.click(screen.getByTestId("submit"));
```
Await user events. Not awaiting them causes `act` errors

nerijus-muranovas commented on src/components/AttendantForm.test.jsx:
```diff
@@ -0,0 +1,151 @@
+      isAttendantsLoading: false,
+      isJobTitlesLoading: false,
+      submitAttendant: jest.fn(),
+      jobTitles: jobTitles,
```
job titles can be defined inline here. In general - if the variable is used once, it is worth considering defining it inline

nerijus-muranovas commented on src/components/AttendantForm.test.jsx:
```diff
@@ -0,0 +1,151 @@
+    render(
+      <AttendantForm
+        {...props}
+        jobTitlesApiError={[ErrorMessages.FAILED_TO_GET_JOB_TITLES]}
```
I'm guessing the array is unintentional

nerijus-muranovas commented on src/components/AttendantForm.test.jsx:
```diff
@@ -0,0 +1,151 @@
+    );
+
+    expect(
+      screen.getByText(ErrorMessages.FAILED_TO_GET_JOB_TITLES)
```
For the assertions use actual text. It is not right/wrong to do it your way, in general in vinted literal text is asserted

nerijus-muranovas commented on src/components/AttendantForm.test.jsx:
```diff
@@ -0,0 +1,151 @@
+      />
+    );
+
+    expect(screen.queryByTestId("job-title")).not.toBeInTheDocument();
```
About test ids - react testing library suggests writing tests that reflect how the user would use the app which implies using querying by roles, labels, text (everything except test ids).

Test ids are fine for some use cases (there are even some performance benefits to doing so especially good for marketplace monolith) and for the sake of this exercise I would suggest trying the former (non-test id) method, to get the hang of it.

agnijaravdo replied:
Good point! Tried to use other selectors as labels, texts, placehoderText for unit tests.
I just can not figure out how to select `job-title` dropdown option without using test-id. Tried by text and by role but no luck 🤔

nerijus-muranovas commented on src/components/AttendantForm.test.jsx:
```diff
@@ -0,0 +1,151 @@
+
+    userEvent.click(screen.getByTestId("age"));
+    userEvent.click(screen.getByTestId("last-name"));
+    await waitFor(() => {
```
waitFor can be removed if userEvents are `await`ed

agnijaravdo committed:
fix formatting with prettier

agnijaravdo committed:
add suggested changes after latest review 1/2

agnijaravdo committed:
use other Jest selector methods instead of selecting only by test-id
